pennylane/transforms/decompositions/single_qubit_unitary.py

```diff
@@ -81,27 +78,24 @@ def zyz_decomposition(U, wire):
     """
     U = _convert_to_su2(U)
 
-    # Check if the matrix is diagonal; only need to check one corner.
-    # If it is diagonal, we don't need a full Rot, just return an RZ.
-    if math.allclose(U[0, 1], [0.0]):
-        omega = 2 * math.angle(U[1, 1])
-        return [qml.RZ(omega, wires=wire)]
-
-    # If the top left element is 0, can only use the off-diagonal elements. We
-    # have to be very careful with the math here to ensure things that get
-    # multiplied together are of the correct type in the different interfaces.
-    if math.allclose(U[0, 0], [0.0]):
-        phi = 0.0
-        theta = -np.pi
-        omega = 1j * math.log(U[0, 1] / U[1, 0]) - np.pi
-    else:
-        # If not diagonal, compute the angle of the RY
-        cos2_theta_over_2 = math.abs(U[0, 0] * U[1, 1])
-        theta = 2 * math.arccos(math.sqrt(cos2_theta_over_2))
-
-        el_division = U[0, 0] / U[1, 0]
-        tan_part = math.cast_like(math.tan(theta / 2), el_division)
-        omega = 1j * math.log(tan_part * el_division)
-        phi = -omega - math.cast_like(2 * math.angle(U[0, 0]), omega)
-
-    return [qml.Rot(math.real(phi), math.real(theta), math.real(omega), wires=wire)]
+    # If the value of U is not abstract, we can include a conditional statement
+    # that will check if the off-diagonal elements are 0; if so, just use one RZ
+    if not math.is_abstract(U):
+        if math.allclose(U[0, 1], 0.0):
+            return [qml.RZ(2 * math.angle(U[1, 1]), wires=wire)]
+
+    # Derive theta from the off-diagonal element. Clip to ensure valid arcsin input
+    element = math.clip(math.abs(U[0, 1]), 0, 1)
+    theta = 2 * math.arcsin(element)
+
+    # Compute phi and omega from the angles of the top row; use atan2 to keep
+    # the angle within -np.pi and np.pi, and add very small values to avoid the
+    # undefined case of 0/0. We add a smaller value to the imaginary part than
+    # the real part because it is imag / real in the definition of atan2.
+    angle_U00 = math.arctan2(math.imag(U[0, 0]) + 1e-128, math.real(U[0, 0]) + 1e-64)
+    angle_U10 = math.arctan2(math.imag(U[1, 0]) + 1e-128, math.real(U[1, 0]) + 1e-64)
```

Comment:
I was reviewing the previous version! Trying to tackle the issue with `cond`. The code is now conditional free and so the problem with `cond` is solved. Does this have any effect on the performance?

Reply:
Yes, actually some pretty nice improvements! So just in the non-jitted version with NumPy, here are some timing results for the new single-qubit decomposition (the new versions are the ones called from the transforms module): 
![image](https://user-images.githubusercontent.com/2068515/156834129-f3887546-3155-4cf7-a74a-8f4d5c92756e.png)

And also for the two-qubit decomposition, which even though no conditionals were removed, still saw some benefits due to improvements in the single-qubit decomp. 

![image](https://user-images.githubusercontent.com/2068515/156834247-a24faeb1-9ca1-4f5a-924c-76d01562f4e8.png)

Reply:
Awesome! :heart_eyes: